Index: qt6-base/src/plugins/platforms/xcb/qxcbdrag.cpp
===================================================================
--- qt6-base.orig/src/plugins/platforms/xcb/qxcbdrag.cpp
+++ qt6-base/src/plugins/platforms/xcb/qxcbdrag.cpp
@@ -860,7 +860,7 @@ void QXcbDrag::handle_xdnd_status(const
     if (event->data.data32[0] && event->data.data32[0] != current_target)
         return;
 
-    const bool dropPossible = event->data.data32[1];
+    const bool dropPossible = event->data.data32[1] & 1;
     setCanDrop(dropPossible);
 
     if (dropPossible) {
