Author: Tian Shilin <tianshilin@uniontech.com>
Date:   Fri Jun 26 13:44:42 2026
Subject: QDom: Change default InvalidDataPolicy to ReturnNullNode
Upstream: https://codereview.qt-project.org/c/qt/qtbase/+/748323


---
Index: qt6-base/src/xml/dom/qdom.cpp
===================================================================
--- qt6-base.orig/src/xml/dom/qdom.cpp
+++ qt6-base/src/xml/dom/qdom.cpp
@@ -91,7 +91,7 @@ static void qt_split_namespace(QString&
  *
  **************************************************************/
 QDomImplementation::InvalidDataPolicy QDomImplementationPrivate::invalidDataPolicy
-    = QDomImplementation::AcceptInvalidChars;
+    = QDomImplementation::ReturnNullNode;
 
 // [5] Name ::= (Letter | '_' | ':') (NameChar)*
 
@@ -541,11 +541,12 @@ bool QDomImplementation::isNull()
     This enum specifies what should be done when a factory function
     in QDomDocument is called with invalid data.
     \value AcceptInvalidChars The data should be stored in the DOM object
-        anyway. In this case the resulting XML document might not be well-formed.
-        This is the default value and QDom's behavior in Qt < 4.1.
+           anyway. In this case the resulting XML document might not be well-formed.
+           This was the default value and QDom's behavior prior to Qt 6.12.
     \value DropInvalidChars The invalid characters should be removed from
-        the data.
+           the data.
     \value ReturnNullNode The factory function should return a null node.
+           This is the default value since Qt 6.12.
 
     \sa setInvalidDataPolicy(), invalidDataPolicy()
 */
